Early implementation of "implementation-only imports"

jrose-apple opened the pull request:
I'm going to implement them with a terrible underscored name and no correctness checking for now.

rdar://problem/48985979

jrose-apple committed:
Change ModuleDecl::getImportedModules to take an option set

...in preparation for me adding a third kind of import, making the
existing "All" kind a problem. NFC, except that I did rewrite the
ClangModuleUnit implementation of getImportedModules to be simpler!

jrose-apple commented:
(at this point I don't even have the implementation, just a refactoring I want to test against LLDB)

jrose-apple commented:
apple/swift-lldb#1139

@swift-ci Please test

jrose-apple commented:
Timeout in an LLDB test -> probably not related. Clear to move forward!

jrose-apple commented:
apple/swift-lldb#1139

@swift-ci Please test

jrose-apple commented:
TODO list of things I can think of off the top of my head:

- LLDB support (debugging FooKit vs. debugging a client of FooKit)
- Restrictions on use:
  - inlinable code, including default arguments
  - types in public and `@usableFromInline` entities
  - stored properties in non-resilient types (including classes) – we probably need a better story for this
- Conformances need to be allowed but not cause problems when we try to deserialize them. (That's currently not the case because we assume conformances might be needed to form the types of stored properties.)
  - This one gets easier if we limit implementation-only imports to only support textual interfaces (because we could just not print them), but textual interfaces require resilience, and that means they can't be used elsewhere, which would be unfortunate.
- How does this interact with `@testable import`?
- Needs more tests.
  - Needs a test for when a module uses `@_uncheckedImplementationOnly` in one file and nothing or `@_exported` in another.

I'm planning to write up a pseudo-proposal soon, in preparation for possibly handing off this feature to someone else.

jrose-apple commented:
Just rebasing, still need to write tests. John's owning this now but I can help out by making sure this is in shape.

@swift-ci Please test

jrose-apple commented:
Ugh. I thought I copied the LLDB PR too.

apple/swift-lldb#1139
@swift-ci Please test

jrose-apple commented:
Okay, I think this is in a reviewable state now. Most of the TODOs still apply, but it has tests, at least!

apple/swift-lldb#1139
@swift-ci Please test

jrose-apple commented:
apple/swift-lldb#1139
@swift-ci Please test macOS

jrose-apple commented:
apple/swift-lldb#1139
@swift-ci Please test macOS

jrose-apple commented:
@brentdax, would you remind reviewing this too? Both so it can get merged sooner rather than later, and to increase the number of people who understand this feature. (That means ask questions if something doesn't make sense.)

I do recommend doing the two commits separately; the first is a lot of updating call sites.

jrose-apple commented:
apple/swift-lldb#1139
@swift-ci Please test

beccadax reviewed (approved):
I have a few suggestions for style and clarity, but they're all suggestions—nothing mandatory.

jrose-apple commented on test/NameBinding/import-implementation-only.swift:
```diff
@@ -0,0 +1,14 @@
+// RUN: %target-swift-frontend -emit-module -o %t/abcde.swiftmodule %S/Inputs/abcde.swift
+// RUN: %target-swift-frontend -emit-module -o %t/Library.swiftmodule %s -I %t
+
+// RUN: echo 'import Library; foo()' > %t/main.swift
```
Gotta save inodes!

More seriously, I don't like having to jump around between files to find out what a test is doing, so if something really is this small I'd rather have it inline.

jrose-apple commented on include/swift/Basic/OptionSet.h:
```diff
@@ -87,6 +87,16 @@ class OptionSet {
   }
 
+  /// Check if two option sets contain the exact same options.
+  friend bool operator==(OptionSet lhs, OptionSet rhs) {
```
:-/ Yeah, I guess so.

jrose-apple committed:
Allow implementation-only imports to be missing when debugging

This will be tested on the LLDB side.

jrose-apple committed:
Add @_implementationOnly

This is an attribute that gets put on an import in library FooKit to
keep it from being a requirement to import FooKit. It's not checked at
all, meaning that in this form it is up to the author of FooKit to
make sure nothing in its API or ABI depends on the implementation-only
dependency. There's also no debugging support here (debugging FooKit
/should/ import the implementation-only dependency if it's present).

The goal is to get to a point where it /can/ be checked, i.e. FooKit
developers are prevented from writing code that would rely on FooKit's
implementation-only dependency being present when compiling clients of
FooKit. But right now it's not.

rdar://problem/48985979

jrose-apple committed:
[Serialization] Preserve @_implementationOnly through module merging

When we build incrementally, we produce "partial swiftmodules" for
each input source file, then merge them together into the final
compiled module that, among other things, gets used for debugging.
Without this, we'd drop @_implementationOnly imports and any types
from the modules that were imported during the module-merging step
and then be unable to debug those types

jrose-apple commented:
[Full macOS build here](https://ci.swift.org/job/swift-PR-osx/12103/), [full Linux tests passed](https://ci.swift.org/job/swift-PR-Linux/12104/)

Updated for Brent's feedback, and also a tiny bit of debugging support (enough for the test cases in apple/swift-lldb#1414).

jrose-apple commented:
apple/swift-lldb#1139
@swift-ci Please smoke test

beccadax reviewed (approved):
Looks great now!

jrose-apple commented:
Hm, I don't *think* those LLDB failures would come from this PR, but maybe?

apple/swift-lldb#1139
@swift-ci Please test macOS

jrose-apple commented:
Why am I not seeing these LLDB failures locally?

jrose-apple commented:
Maybe…

apple/swift-lldb#1139
@swift-ci Please clean test macOS

jrose-apple commented:
\*sigh\* Well, I'm not sure whether the clean test did it, or whether something just got fixed (or disabled) on master, but it looks like we're clear to move forward. Merging.
